File: ai-env-book/build-docx.py
#!/usr/bin/env python3
"""Build DOCX for Your Dev Environment: A Guide for AI-Assisted Developers."""

from pathlib import Path
import re

from docx import Document
from docx.enum.section import WD_SECTION
from docx.enum.style import WD_STYLE_TYPE
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.shared import Inches, Pt, RGBColor

BOOK_DIR = Path(__file__).parent
MD_PATH = BOOK_DIR / "your-dev-environment.md"
COVER_PATH = BOOK_DIR / "DevEnvCover.png"
OUT_PATH = BOOK_DIR / "your-dev-environment.docx"

TITLE = "Your Dev Environment"
SUBTITLE = "A Guide for AI-Assisted Developers"
AUTHOR = "Lee Harrington"

DARK = RGBColor(0x1A, 0x1A, 0x2E)
MID = RGBColor(0x2E, 0x40, 0x57)
GREY = RGBColor(0x55, 0x55, 0x55)


def set_page_layout(doc: Document) -> None:
    for section in doc.sections:
        section.page_width = Inches(6)
        section.page_height = Inches(9)
        section.left_margin = Inches(0.7)
        section.right_margin = Inches(0.7)
        section.top_margin = Inches(0.7)
        section.bottom_margin = Inches(0.7)


def ensure_code_style(doc: Document) -> None:
    normal = doc.styles["Normal"]
    normal.font.name = "Georgia"
    normal.font.size = Pt(11)

    h1 = doc.styles["Heading 1"]
    h1.font.name = "Arial"
    h1.font.size = Pt(18)
    h1.font.bold = True
    h1.font.color.rgb = DARK

    h2 = doc.styles["Heading 2"]
    h2.font.name = "Arial"
    h2.font.size = Pt(14)
    h2.font.bold = True
    h2.font.color.rgb = MID

    h3 = doc.styles["Heading 3"]
    h3.font.name = "Arial"
    h3.font.size = Pt(12)
    h3.font.bold = True

    if "CodeBlock" not in [style.name for style in doc.styles]:
        style = doc.styles.add_style("CodeBlock", WD_STYLE_TYPE.PARAGRAPH)
        style.font.name = "Courier New"
        style.font.size = Pt(9)
        style.paragraph_format.left_indent = Inches(0.3)
        style.paragraph_format.space_before = Pt(4)
        style.paragraph_format.space_after = Pt(4)


def add_title_pages(doc: Document) -> None:
    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    p.add_run().add_picture(str(COVER_PATH), width=Inches(5.4))

    doc.add_section(WD_SECTION.NEW_PAGE)

    for _ in range(3):
        doc.add_paragraph()

    title = doc.add_paragraph()
    title.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = title.add_run(TITLE)
    run.font.name = "Arial"
    run.font.size = Pt(24)
    run.bold = True
    run.font.color.rgb = DARK

    subtitle = doc.add_paragraph()
    subtitle.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = subtitle.add_run(SUBTITLE)
    run.font.name = "Arial"
    run.font.size = Pt(13)
    run.italic = True
    run.font.color.rgb = GREY

    author = doc.add_paragraph()
    author.alignment = WD_ALIGN_PARAGRAPH.CENTER
    author.paragraph_format.space_before = Pt(20)
    run = author.add_run(AUTHOR)
    run.font.name = "Arial"
    run.font.size = Pt(15)
    run.font.color.rgb = MID

    doc.add_section(WD_SECTION.NEW_PAGE)


def add_inline_runs(paragraph, text: str) -> None:
    pattern = re.compile(r"(`[^`]+`|\*\*[^*]+\*\*|\*[^*]+\*)")
    pos = 0
    for match in pattern.finditer(text):
        if match.start() > pos:
            paragraph.add_run(text[pos:match.start()])
        token = match.group(0)
        if token.startswith("`"):
            run = paragraph.add_run(token[1:-1])
            run.font.name = "Courier New"
        elif token.startswith("**"):
            run = paragraph.add_run(token[2:-2])
            run.bold = True
        else:
            run = paragraph.add_run(token[1:-1])
            run.italic = True
        pos = match.end()
    if pos < len(text):
        paragraph.add_run(text[pos:])


def render_markdown(doc: Document, md_text: str) -> None:
    lines = md_text.splitlines()
    in_code = False
    code_lines = []
    in_callout = False
    callout_lines = []

    def flush_code() -> None:
        nonlocal code_lines
        if not code_lines:
            return
        for line in code_lines:
            doc.add_paragraph(line if line else " ", style="CodeBlock")
        code_lines = []

    def flush_callout() -> None:
        nonlocal callout_lines
        if not callout_lines:
            return
        for raw in callout_lines:
            paragraph = doc.add_paragraph()
            paragraph.paragraph_format.left_indent = Inches(0.3)
            paragraph.paragraph_format.space_before = Pt(3)
            paragraph.paragraph_format.space_after = Pt(3)
            add_inline_runs(paragraph, re.sub(r"^>\s*", "", raw))
        callout_lines = []

    for line in lines:
        if line.startswith("```"):
            if in_code:
                flush_code()
                in_code = False
            else:
                flush_callout()
                in_code = True
            continue

        if in_code:
            code_lines.append(line)
            continue

        if line.startswith(">"):
            in_callout = True
            callout_lines.append(line)
            continue

        if in_callout:
            flush_callout()
            in_callout = False

        stripped = line.strip()
        if not stripped:
            continue

        if stripped == "---":
            doc.add_paragraph()
            continue

        if stripped.startswith("# "):
            p = doc.add_paragraph(style="Heading 1")
            add_inline_runs(p, stripped[2:])
            continue
        if stripped.startswith("## "):
            p = doc.add_paragraph(style="Heading 2")
            add_inline_runs(p, stripped[3:])
            continue
        if stripped.startswith("### "):
            p = doc.add_paragraph(style="Heading 3")
            add_inline_runs(p, stripped[4:])
            continue

        if re.match(r"^\d+\. ", stripped):
            p = doc.add_paragraph(style="List Number")
            add_inline_runs(p, re.sub(r"^\d+\. ", "", stripped))
            continue

        if stripped.startswith("- "):
            p = doc.add_paragraph(style="List Bullet")
            add_inline_runs(p, stripped[2:])
            continue

        p = doc.add_paragraph()
        p.paragraph_format.space_after = Pt(8)
        add_inline_runs(p, stripped)

    if in_callout:
        flush_callout()
    if in_code:
        flush_code()


def main() -> None:
    doc = Document()
    set_page_layout(doc)
    ensure_code_style(doc)
    core = doc.core_properties
    core.title = f"{TITLE}: {SUBTITLE}"
    core.author = AUTHOR
    core.subject = "AI-assisted development environment guide"
    core.comments = "Generated from your-dev-environment.md"
    add_title_pages(doc)
    render_markdown(doc, MD_PATH.read_text(encoding="utf-8"))
    doc.save(str(OUT_PATH))
    size = OUT_PATH.stat().st_size
    print(f"Written: {OUT_PATH} ({size // 1024}KB)")


if __name__ == "__main__":
    main()

File: ai-env-book/build-docx.js
#!/usr/bin/env node
/**
 * Build script: your-dev-environment.docx
 * Converts your-dev-environment.md → formatted Word document with cover image
 */

const {
  Document, Packer, Paragraph, TextRun, Table, TableRow, TableCell, ImageRun,
  Header, Footer, AlignmentType, HeadingLevel, BorderStyle, WidthType, ShadingType,
  PageNumber, PageBreak, LevelFormat, TableOfContents, SectionType
} = require('docx');
const fs = require('fs');
const path = require('path');

const BOOK_DIR  = path.dirname(__filename);
const MD_PATH   = path.join(BOOK_DIR, 'your-dev-environment.md');
const COVER_PATH = path.join(BOOK_DIR, 'DevEnvCover.png');
const OUT_PATH  = path.join(BOOK_DIR, 'your-dev-environment.docx');

// ─── Page dimensions (6×9 trade paperback) ──────────────────────────────────
const PAGE_W    = 8640;   // 6 inches
const PAGE_H    = 12960;  // 9 inches
const MARGIN    = 1008;   // 0.7 inch
const CONTENT_W = PAGE_W - MARGIN * 2;

// ─── Colours ─────────────────────────────────────────────────────────────────
const BRAND_DARK = '1A1A2E';
const BRAND_MID  = '2E4057';
const CALLOUT_BG = 'F0F4F8';
const CODE_BG    = 'F5F5F5';
const GREEN      = '2E7D32';

// ─── Helpers ─────────────────────────────────────────────────────────────────
function spacer(pts = 6) {
  return new Paragraph({ children: [new TextRun('')], spacing: { before: 0, after: pts * 20 } });
}

function hrule() {
  return new Paragraph({
    children: [new TextRun('')],
    border: { bottom: { style: BorderStyle.SINGLE, size: 4, color: 'CCCCCC', space: 6 } },
    spacing: { before: 120, after: 120 },
  });
}

function parseInline(text) {
  const runs = [];
  let remaining = text.replace(/<!--[\s\S]*?-->/g, '').trim();
  while (remaining.length > 0) {
    const codeMatch = remaining.match(/^`([^`]+)`/);
    if (codeMatch) {
      runs.push(new TextRun({ text: codeMatch[1], font: 'Courier New', size: 18, color: '333333', shading: { fill: CODE_BG, type: ShadingType.CLEAR } }));
      remaining = remaining.slice(codeMatch[0].length); continue;
    }
    const boldMatch = remaining.match(/^\*\*([^*]+)\*\*/);
    if (boldMatch) {
      runs.push(new TextRun({ text: boldMatch[1], bold: true }));
      remaining = remaining.slice(boldMatch[0].length); continue;
    }
    const italicMatch = remaining.match(/^\*([^*]+)\*/) || remaining.match(/^_([^_]+)_/);
    if (italicMatch) {
      runs.push(new TextRun({ text: italicMatch[1], italics: true }));
      remaining = remaining.slice(italicMatch[0].length); continue;
    }
    const next = remaining.search(/[`*_]/);
    if (next === -1) { runs.push(new TextRun(remaining)); remaining = ''; }
    else { runs.push(new TextRun(remaining.slice(0, next))); remaining = remaining.slice(next); }
  }
  return runs.length ? runs : [new TextRun(text)];
}

function calloutParagraph(lines) {
  const label = lines[0].replace(/^>\s*\*\*([^*]+)\*\*.*/, '$1');
  const color = label.includes('Key Takeaway') ? GREEN
              : label.includes('Watch')        ? BRAND_MID
              : label.includes("Don't")        ? 'B71C1C'
              : BRAND_MID;
  return lines.map((line, i) => new Paragraph({
    children: parseInline(line.replace(/^>\s*/, '')),
    indent: { left: 360, right: 360 },
    spacing: { before: i === 0 ? 120 : 60, after: 60 },
    shading: { fill: CALLOUT_BG, type: ShadingType.CLEAR },
    border: { left: { style: BorderStyle.SINGLE, size: 12, color, space: 6 } },
  }));
}

function codeBlock(lines) {
  return lines.map((line, i) => new Paragraph({
    children: [new TextRun({ text: line.length ? line : ' ', font: 'Courier New', size: 18, color: '333333' })],
    shading: { fill: CODE_BG, type: ShadingType.CLEAR },
    indent: { left: 360, right: 360 },
    spacing: { before: i === 0 ? 120 : 0, after: i === lines.length - 1 ? 120 : 0, line: 240 },
    border: {
      top:    i === 0              ? { style: BorderStyle.SINGLE, size: 2, color: 'DDDDDD' } : undefined,
      bottom: i === lines.length-1 ? { style: BorderStyle.SINGLE, size: 2, color: 'DDDDDD' } : undefined,
      left:   { style: BorderStyle.SINGLE, size: 2, color: 'DDDDDD' },
      right:  { style: BorderStyle.SINGLE, size: 2, color: 'DDDDDD' },
    }
  }));
}

function parseMarkdown(mdText) {
  const lines = mdText.split('\n');
  const paragraphs = [];
  let i = 0, inCode = false, inCallout = false;
  let codeLines = [], calloutLines = [];

  const flushCallout = () => {
    if (!calloutLines.length) return;
    paragraphs.push(...calloutParagraph(calloutLines), spacer(4));
    calloutLines = []; inCallout = false;
  };

  while (i < lines.length) {
    const line = lines[i];

    if (line.startsWith('```')) {
      if (!inCode) { flushCallout(); inCode = true; codeLines = []; }
      else { paragraphs.push(...codeBlock(codeLines)); inCode = false; codeLines = []; }
      i++; continue;
    }
    if (inCode) { codeLines.push(line); i++; continue; }

    if (line.startsWith('>')) {
      inCallout = true; calloutLines.push(line); i++; continue;
    }
    if (inCallout && line.trim() === '') {
      let j = i + 1;
      while (j < lines.length && lines[j].trim() === '') j++;
      if (j < lines.length && lines[j].startsWith('>')) { calloutLines.push(''); i++; continue; }
      flushCallout(); i++; continue;
    }
    if (inCallout) flushCallout();

    if (line.startsWith('<!--')) { while (i < lines.length && !lines[i].includes('-->')) i++; i++; continue; }
    if (/^---+$/.test(line.trim())) { paragraphs.push(hrule()); i++; continue; }

    if (line.startsWith('# ')) {
      paragraphs.push(new Paragraph({
        children: parseInline(line.slice(2).trim()),
        heading: HeadingLevel.HEADING_1,
        pageBreakBefore: true,
        spacing: { before: 480, after: 240 },
      }));
      i++; continue;
    }
    if (line.startsWith('## ')) {
      paragraphs.push(new Paragraph({
        children: parseInline(line.slice(3).trim()),
        heading: HeadingLevel.HEADING_2,
        spacing: { before: 360, after: 160 },
      }));
      i++; continue;
    }
    if (line.startsWith('### ')) {
      paragraphs.push(new Paragraph({
        children: parseInline(line.slice(4).trim()),
        heading: HeadingLevel.HEADING_3,
        spacing: { before: 240, after: 120 },
      }));
      i++; continue;
    }

    if (/^[-*] /.test(line)) {
      paragraphs.push(new Paragraph({
        children: parseInline(line.slice(2).trim()),
        numbering: { reference: 'bullets', level: 0 },
        spacing: { before: 40, after: 40 },
      }));
      i++; continue;
    }
    if (/^\d+\. /.test(line)) {
      paragraphs.push(new Paragraph({
        children: parseInline(line.replace(/^\d+\. /, '').trim()),
        numbering: { reference: 'numbers', level: 0 },
        spacing: { before: 40, after: 40 },
      }));
      i++; continue;
    }

    // table row — simplified: render as plain paragraphs
    if (line.startsWith('|')) {
      if (/^\|[-| ]+\|$/.test(line.trim())) { i++; continue; } // skip separator
      const cells = line.split('|').filter((c, idx) => idx > 0 && idx < line.split('|').length - 1).map(c => c.trim());
      paragraphs.push(new Paragraph({
        children: cells.flatMap((c, ci) => [
          ...(ci > 0 ? [new TextRun('  |  ')] : []),
          ...parseInline(c),
        ]),
        spacing: { before: 40, after: 40 },
      }));
      i++; continue;
    }

    if (line.trim() === '') { i++; continue; }

    paragraphs.push(new Paragraph({
      children: parseInline(line.trim()),
      spacing: { before: 0, after: 160, line: 276 },
    }));
    i++;
  }
  if (inCallout) flushCallout();
  if (inCode) paragraphs.push(...codeBlock(codeLines));
  return paragraphs;
}

async function main() {
  const mdText   = fs.readFileSync(MD_PATH, 'utf8');
  const coverImg = fs.readFileSync(COVER_PATH);
  const body     = parseMarkdown(mdText);

  const pageProps = {
    page: {
      size: { width: PAGE_W, height: PAGE_H },
      margin: { top: MARGIN, right: MARGIN, bottom: MARGIN, left: MARGIN },
    },
  };

  const doc = new Document({
    numbering: {
      config: [
        { reference: 'bullets', levels: [{ level: 0, format: LevelFormat.BULLET, text: '\u2022', alignment: AlignmentType.LEFT, style: { paragraph: { indent: { left: 720, hanging: 360 } } } }] },
        { reference: 'numbers', levels: [{ level: 0, format: LevelFormat.DECIMAL, text: '%1.', alignment: AlignmentType.LEFT, style: { paragraph: { indent: { left: 720, hanging: 360 } } } }] },
      ],
    },
    styles: {
      default: { document: { run: { font: 'Georgia', size: 22 } } },
      paragraphStyles: [
        { id: 'Heading1', name: 'Heading 1', basedOn: 'Normal', next: 'Normal', quickFormat: true,
          run: { font: 'Arial', size: 36, bold: true, color: BRAND_DARK },
          paragraph: { spacing: { before: 480, after: 240 }, outlineLevel: 0 } },
        { id: 'Heading2', name: 'Heading 2', basedOn: 'Normal', next: 'Normal', quickFormat: true,
          run: { font: 'Arial', size: 28, bold: true, color: BRAND_MID },
          paragraph: { spacing: { before: 360, after: 160 }, outlineLevel: 1 } },
        { id: 'Heading3', name: 'Heading 3', basedOn: 'Normal', next: 'Normal', quickFormat: true,
          run: { font: 'Arial', size: 24, bold: true, color: '444444' },
          paragraph: { spacing: { before: 240, after: 120 }, outlineLevel: 2 } },
      ],
    },
    sections: [
      // Cover
      {
        properties: { ...pageProps, type: SectionType.NEXT_PAGE },
        children: [new Paragraph({
          children: [new ImageRun({
            type: 'png',
            data: coverImg,
            transformation: { width: 540, height: 720 },
            altText: { title: 'Cover', description: 'Book cover', name: 'cover' },
          })],
          alignment: AlignmentType.CENTER,
          spacing: { before: 0, after: 0 },
        })],
      },
      // Title page
      {
        properties: { ...pageProps, type: SectionType.NEXT_PAGE },
        children: [
          spacer(120),
          new Paragraph({ children: [new TextRun({ text: 'Your Dev Environment', font: 'Arial', size: 52, bold: true, color: BRAND_DARK })], alignment: AlignmentType.CENTER, spacing: { before: 0, after: 180 } }),
          new Paragraph({ children: [new TextRun({ text: 'A Guide for AI-Assisted Developers', font: 'Arial', size: 26, italics: true, color: '555555' })], alignment: AlignmentType.CENTER, spacing: { before: 0, after: 720 } }),
          hrule(),
          spacer(60),
          new Paragraph({ children: [new TextRun({ text: 'Lee Harrington', font: 'Arial', size: 32, color: BRAND_MID })], alignment: AlignmentType.CENTER, spacing: { before: 240, after: 120 } }),
        ],
      },
      // TOC
      {
        properties: { ...pageProps, type: SectionType.NEXT_PAGE },
        children: [
          new Paragraph({ children: [new TextRun({ text: 'Table of Contents', font: 'Arial', size: 36, bold: true, color: BRAND_DARK })], heading: HeadingLevel.HEADING_1, spacing: { before: 0, after: 360 } }),
          new TableOfContents('Table of Contents', { hyperlink: true, headingStyleRange: '1-2' }),
        ],
      },
      // Body
      {
        properties: { ...pageProps, type: SectionType.NEXT_PAGE },
        footers: {
          default: new Footer({ children: [new Paragraph({
            children: [
              new TextRun({ text: 'Your Dev Environment  •  ', color: '888888', size: 18 }),
              new TextRun({ children: [PageNumber.CURRENT], color: '888888', size: 18 }),
            ],
            alignment: AlignmentType.CENTER,
          })] }),
        },
        children: body,
      },
    ],
  });

  const buffer = await Packer.toBuffer(doc);
  fs.writeFileSync(OUT_PATH, buffer);
  console.log(`Written: ${OUT_PATH} (${Math.round(buffer.length / 1024)}KB)`);
}

main().catch(err => { console.error(err); process.exit(1); });
